src/Http/Http/src/Features/RequestServicesFeature.cs

```diff
@@ -28,7 +24,7 @@ public IServiceProvider RequestServices
         {
             get
             {
-                if (!_requestServicesSet)
+                if (!_requestServicesSet && _scopeFactory != null)
```

Comment:
When is _scopeFactory null now?

If for some reason, we can no longer assert _scopeFactory is not null in the RequestServicesFeature ctor, we should at least throw from the RequestServices getter if both _requestServices and _scopeFactory are not set.

By returning null, we're just moving the NullReferenceException into user code with no explanation for why RequestServices was null to begin with. Previously, RequestServices could never be null before unless you explicitly set it to null beforehand, so I doubt anyone is checking whether RequestServices is null before using it.

Reply:
I saw the following in the PR description:

> If you derived from HttpContextFactory and used one of the old constructors, then RequestServices will be broken.

If that's the case we should just break as early as possible and not only obsolete but also throw a NotSupportedException from the old constructors that completely break RequestServices.

Reply:
It’s described in the PR description. If you derived from the HttpContextFactory and use one of the older constructors, it’ll likely start blowing up with a null ref. I don’t have a great way to mitigate that that’s why the type is now obsolete.

I can throw in the older constructor to make it more obvious.

Reply:
Yes that effectively makes the the older constructors completely useless.

Reply:
Done!

Reply:
I get not throwing in the ctor, but you could have at least thrown a useful error here instead of returning null when `!_requestServicesSet && _scopeFactory == null`. Otherwise, this is just moving the NullReferenceException.